### PR TITLE
*:initial commit.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: select.proto [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+// KeyPoint represents a single key.
+message KeyPoint {
+	optional bytes key = 1;
+}
+
+// SelectRequest works like a simplified select statement.
+message SelectRequest {
+	// Only read data commited before this timestamp.
+	optional int64 timestamp = 1;
```
Comment (ngaut) [inline select.proto]: s/timestamp/start_ts

File: select.proto [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	optional int64 scan_limit = 9;
+	optional int64 result_limit = 10;
+}
+
+// Row packs all column values into a single byte slice.
+// With lengths to separate them.
+// values are all in text format.
+message Row {
+	repeated int32 lengths = 1;
+	optional bytes values = 2;
```
Comment (shenli) [inline select.proto]: repeated bytes values = 1; is  better?

File: select.proto [hunk 3]
```diff
@@ -0,0 +1,76 @@
+syntax = "proto2";
+
```
Comment (nieyy) [inline select.proto]:
```
option java_multiple_files = true;
option java_package = "com.pingcap.tidb.tipb";
option java_outer_classname = "SelectProto";
```

File: select.proto [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+// Row packs all column values into a single byte slice.
+// With lengths to separate them.
+// values are all in text format.
+message Row {
+	repeated int32 lengths = 1;
+	optional bytes values = 2;
+}
+
+// Response for SelectRequest.
+message SelectResponse {
```
Comment (nieyy) [inline select.proto]: Can we just use `SelectRequest` and `SelectResponse` to represent the pushdown information of all kinds of `SELECT` statements?

**Comment by nieyy** [comment]
The name and definition of the .proto files should have an obvious relationship related to KV capability interfaces.

File: select.proto (hunk 4 above)
Comment (coocood) [inline select.proto]: It's not the original SQL `SELECT` statement. It only supports a single table or single index.
For example, a Join SELECT will be splitted into multiple pushdown `SelectReuqest`.

**Comment by coocood** [comment]
@nieyy capability information can be stored in kv driver, it doesn't need to be included in proto file.

**Comment by nieyy** [comment]
@coocood let's see the definition of kv interfaces later.

File: expression.proto [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	Plus = 103;
+	Minus = 104;
+
+	// TODO: add more expression types.
+}
+
+// Evaluators should implement evaluation functions for every expression type.
+message Expression {
+	optional ExpressionType tp = 1;
+	optional bytes val = 2;
```
Comment (shenli) [inline expression.proto]: Need flag?

File: schema.proto [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	latin1_swedish_ci = 8;
+	utf8_general_ci = 33;
+	utf8mb4_general_ci = 45;
+	binary = 63;
+
+	// TODO: add more collations.
+}
+
+message TableInfo {
+	optional int64 table_id = 1;
```
Comment (shenli) [inline schema.proto]: Should we use go style name? table_id -> tableID?

File: schema.proto [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	repeated ColumnInfo columns = 2;
+}
+
+message ColumnInfo {
+	optional int64 column_id = 1;
+	optional MysqlType tp = 2;
```
Comment (shenli) [inline schema.proto]: Define a message FieldType?

File: select.proto [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	repeated KeyPoint points = 6;
+
+	// where condition.
+	optional Expression where = 7;
+
+	// group by clause.
+	optional Expression group_by = 8;
+
+	// scan_limit should be greater than or equal to result_limit;
+	optional int64 scan_limit = 9;
```
Comment (shenli) [inline select.proto]: Can we remove scan_limit?

File: select.proto [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+message Row {
+	repeated bytes values = 1;
+}
+
+// Response for SelectRequest.
+message SelectResponse {
+	optional Error error = 1;
+
+	// Ranges the region server does not have.
+	repeated KeyRange missing_ranges = 2;
```
Comment (shenli) [inline select.proto]: This should be handled by tikv client. I think SQL layer should not consider this.

File: error.proto [hunk 10]
```diff
@@ -0,0 +1,17 @@
+syntax = "proto2";
+
+package tipb;
+
+option java_multiple_files = true;
+option java_package = "com.pingcap.tidb.tipb";
+option java_outer_classname = "SelectProto";
```
Comment (siddontang) [inline error.proto]: no need `option go_package = ""` ?

File: schema.proto [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+	optional Collation collation = 3;
+	optional int32 columnLen = 4;
+	optional int32 decimal = 5;
+	optional bool pk_handle = 6; // PK handle column value is row handle.
+}
+
+message IndexInfo {
+	optional int64 table_id = 1;
+	optional int64 index_id = 2;
+	repeated ColumnInfo columns = 3;
```
Comment (shenli) [inline schema.proto]: column id list is enough?

File: expression.proto [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+option java_outer_classname = "SelectProto";
+
+enum ExpressionType {
+	// children count 0.
+	// values are in text format.
+	Null = 0;
+	Int64 = 1;
+	Float64 = 2;
+	String = 3;
+	ColumnRef = 21;
```
Comment (qiuyesuifeng) [inline expression.proto]: What is the meaning of `ColumnRef`?

File: schema.proto [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+	TypeLongBlob = 0xfb;
+	TypeBlob = 0xfc;
+	TypeVarString = 0xfd;
+	TypeString = 0xfe;
+	TypeGeometry = 0xff;
+}
+
+// Collation has the same constant value as MySQL.
+enum Collation {
+	latin1_swedish_ci = 8;
```
Comment (qiuyesuifeng) [inline schema.proto]: Should we unify variable name format?

File: expression.proto (hunk 12 above)
Comment (coocood) [inline expression.proto]: column reference. value is column id.

File: select.proto [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+// low is closed, high is open ( low <= value < high ).
+message HandleRange {
+	optional int64 low = 1;
+	optional int64 high = 2;
+}
+
+// SelectRequest works like a simplified select statement.
+message SelectRequest {
+	// transaction start timestamp.
+	optional int64 start_ts = 1;
```
Comment (qiuyesuifeng) [inline select.proto]: Can we use ts instead, seems kv need not know the transaction, only using in MVCC?

File: expression.proto (hunk 5 above)
Comment (coocood) [inline expression.proto]: No, flag can be recalculated. 
And in this scenario, it's rarely used.

File: select.proto [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+	optional bytes low = 1;
+	optional bytes high = 2;
+}
+
+message IndexRequest {
+	// transaction start timestamp.
+	optional int64 start_ts = 1;
+
+	// If IndexInfo is not null, table_info would be null.
+	optional IndexInfo index_info = 2;
```
Comment (qiuyesuifeng) [inline select.proto]: Where is the table_info, seems no such a variable in IndexRequest?

File: select.proto [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+	repeated bytes points = 4;
+
+	optional int64 limit = 5;
+}
+
+// IndexResponse is the
+message IndexResponse {
+	optional Error error = 1;
+	repeated int64 handles = 2;
+	optional bytes next_key = 3;
```
Comment (qiuyesuifeng) [inline select.proto]: Is it better to use a boolean variable to indicate that there is more data to be read? Seems that is much clearer.

File: schema.proto (hunk 11 above)
Comment (coocood) [inline schema.proto]: We may need to use column info to decode index value.

File: select.proto (hunk 16 above)
Comment (coocood) [inline select.proto]: Then we will have a redundant field. 
Add some comment on it should be enough.

File: error.proto (hunk 10 above)
Comment (coocood) [inline error.proto]: the go package name is package 'tipb'.

**Comment by coocood** [comment]
@nieyy @shenli @qiuyesuifeng @siddontang PTAL

**Comment by BusyJay** [comment]
ping

**Comment by coocood** [comment]
@ngaut PTAL

File: expression.proto [hunk 17]
```diff
@@ -0,0 +1,147 @@
+syntax = "proto2";
```
Comment (ngaut) [inline expression.proto]: Please add license header.

**Comment by shenli** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**